doxygen/man/man3/cMySQLPlugin.3
.TH "cMySQLPlugin" 3 "Thu Feb 11 2016" "Kooky" \" -*- nroff -*-
.ad l
.nh
.SH NAME
cMySQLPlugin \- MySQL Database Connector Plugin class\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <cmysqlplugin\&.h>\fP
.PP
Inherits QObject, and \fBcDBInterface\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "qint16 \fBpluginAPIVersion\fP ()"
.br
.RI "\fIReturns the version of the plugin API\&. \fP"
.ti -1c
.RI "QString \fBpluginName\fP ()"
.br
.RI "\fIReturns the name of the plugin\&. \fP"
.ti -1c
.RI "qint16 \fBpluginVersion\fP ()"
.br
.RI "\fIreturns the version of the plugin \fP"
.ti -1c
.RI "\fBiType\fP \fBpluginType\fP ()"
.br
.RI "\fIreturns the type of the plugin\&. \fP"
.ti -1c
.RI "bool \fBconnect\fP ()"
.br
.RI "\fIconnects to the database\&. \fP"
.ti -1c
.RI "bool \fBinit\fP ()"
.br
.RI "\fIinitializes default settings for the plugin\&. \fP"
.ti -1c
.RI "QString \fBlastError\fP ()"
.br
.RI "\fIreturns the last error string\&. \fP"
.ti -1c
.RI "bool \fBconfig\fP ()"
.br
.RI "\fIcalls the configuration dialog\&. \fP"
.in -1c
.SS "Protected Member Functions"

.in +1c
.ti -1c
.RI "bool \fBopen\fP ()"
.br
.RI "\fIopens the database\&. \fP"
.in -1c
.SS "Protected Attributes"

.in +1c
.ti -1c
.RI "QSqlDatabase \fBm_db\fP"
.br
.ti -1c
.RI "QString \fBm_szLastError\fP"
.br
.in -1c
.SS "Additional Inherited Members"
.SH "Detailed Description"
.PP 
This class implements the database connector for MySQL\&. All functions may be overwritten by derriving classes\&.
.PP
\fBNote:\fP
.RS 4
.RE
.PP
\fBAuthor:\fP
.RS 4
Herwig Birke
.RE
.PP
\fBVersion:\fP
.RS 4
1\&.0
.RE
.PP
\fBDate:\fP
.RS 4
$Date: 2016/02/09 
.RE
.PP

.SH "Member Function Documentation"
.PP 
.SS "bool cMySQLPlugin::config ()\fC [virtual]\fP"

.PP
\fBReturns:\fP
.RS 4
bool true if the configuration has been changed successfully or false otherwise\&. 
.RE
.PP

.PP
Implements \fBcInterface\fP\&.
.PP
References cConfigDialog::database(), cConfigDialog::hostname(), cDBInterface::m_bConnected, m_db, open(), cConfigDialog::password(), pluginName(), cConfigDialog::setDatabase(), cConfigDialog::setHostname(), cConfigDialog::setPassword(), cConfigDialog::setUserName(), and cConfigDialog::userName()\&.
.PP
Referenced by pluginType()\&.
.SS "bool cMySQLPlugin::connect ()\fC [virtual]\fP"

.PP
\fBReturns:\fP
.RS 4
bool true on success or false otherwise 
.RE
.PP

.PP
Implements \fBcDBInterface\fP\&.
.PP
References m_db, open(), and pluginName()\&.
.PP
Referenced by pluginType()\&.
.SS "bool cMySQLPlugin::init ()\fC [virtual]\fP"

.PP
\fBReturns:\fP
.RS 4
bool true if settings where successfully tested or false otherwise\&. 
.RE
.PP

.PP
Implements \fBcDBInterface\fP\&.
.PP
References cDBInterface::m_bConnected, m_db, m_szLastError, and pluginName()\&.
.PP
Referenced by pluginType()\&.
.SS "QString cMySQLPlugin::lastError ()"

.PP
\fBReturns:\fP
.RS 4
QString last error 
.RE
.PP

.PP
References m_szLastError\&.
.PP
Referenced by pluginType()\&.
.SS "bool cMySQLPlugin::open ()\fC [protected]\fP"

.PP
\fBReturns:\fP
.RS 4
bool true on success or false otherwise\&. 
.RE
.PP

.PP
References cDBInterface::m_bConnected, m_db, and m_szLastError\&.
.PP
Referenced by config(), connect(), and pluginType()\&.
.SS "qint16 cMySQLPlugin::pluginAPIVersion ()\fC [inline]\fP, \fC [virtual]\fP"

.PP
\fBReturns:\fP
.RS 4
qint16 plugin API version 
.RE
.PP

.PP
Implements \fBcInterface\fP\&.
.SS "QString cMySQLPlugin::pluginName ()\fC [inline]\fP, \fC [virtual]\fP"

.PP
\fBReturns:\fP
.RS 4
QString name of the plugin 
.RE
.PP

.PP
Implements \fBcInterface\fP\&.
.PP
Referenced by config(), connect(), and init()\&.
.SS "\fBiType\fP cMySQLPlugin::pluginType ()\fC [inline]\fP, \fC [virtual]\fP"

.PP
\fBReturns:\fP
.RS 4
iType type of the plugin\&.
.RE
.PP
Returns the type of the plugin\&. Possible values are: 
.PD 0

.IP "\(bu" 2
\fBcInterface::iTypeDB\fP 
.br
 plugin provides DB connector 
.IP "\(bu" 2
\fBcInterface::iTypeExport\fP 
.br
 plugin provides export capabilities 
.IP "\(bu" 2
\fBcInterface::iTypeImport\fP 
.br
 plugin provides import capabilities  
.PP

.PP
Implements \fBcInterface\fP\&.
.PP
References config(), connect(), init(), cInterface::iTypeDB, lastError(), and open()\&.
.SS "qint16 cMySQLPlugin::pluginVersion ()\fC [inline]\fP, \fC [virtual]\fP"

.PP
\fBReturns:\fP
.RS 4
qint16 version of the plugin 
.RE
.PP

.PP
Implements \fBcInterface\fP\&.
.SH "Member Data Documentation"
.PP 
.SS "QSqlDatabase cMySQLPlugin::m_db\fC [protected]\fP"
holds the database connection 
.PP
Referenced by config(), connect(), init(), and open()\&.
.SS "QString cMySQLPlugin::m_szLastError\fC [protected]\fP"
holds the last error message 
.PP
Referenced by init(), lastError(), and open()\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for Kooky from the source code\&.


dMySQL/cmysqlplugin.cpp
#include "cmysqlplugin.h"
#include "cconfigdialog.h"

#include <QSqlError>
#include <QSettings>
#include <QSqlQuery>
#include <QMessageBox>


bool cMySQLPlugin::connect()
{
	QSettings	settings;

	m_db	= QSqlDatabase::addDatabase("QMYSQL");
	m_db.setHostName(settings.value(QString("plugins/%1/hostName").arg(pluginName())).toString());
	m_db.setDatabaseName(settings.value(QString("plugins/%1/databaseName").arg(pluginName())).toString());
	m_db.setUserName(settings.value(QString("plugins/%1/userName").arg(pluginName())).toString());
	m_db.setPassword(settings.value(QString("plugins/%1/password").arg(pluginName())).toString());

	return(open());
}

bool cMySQLPlugin::init()
{
	QSettings	settings;

	m_db	= QSqlDatabase::addDatabase("QMYSQL");
	m_db.setHostName(settings.value(QString("plugins/%1/hostName").arg(pluginName())).toString());
	m_db.setDatabaseName(settings.value(QString("plugins/%1/databaseName").arg(pluginName())).toString());
	m_db.setUserName(settings.value(QString("plugins/%1/userName").arg(pluginName())).toString());
	m_db.setPassword(settings.value(QString("plugins/%1/password").arg(pluginName())).toString());

	m_bConnected	= m_db.open();

	if(m_bConnected)
	{
		QStringList	tables	= m_db.tables();
		if(tables.count())
		{
			if(QMessageBox::question(0, "Tables exist", "Database not empty.\nExisting tables will be deleted.\nContinue?", QMessageBox::Yes | QMessageBox::No, QMessageBox::No) == QMessageBox::No)
				return(m_bConnected);
		}
		QSqlQuery	query(m_db);
		for(int z = 0;z < tables.count();z++)
			query.exec(QString("DROP TABLE %1;").arg(tables.at(z)));

		if(!query.exec("CREATE TABLE `kooky`.settings ( `var` VARCHAR(100) NOT NULL , `val` VARCHAR(100) NOT NULL ) ENGINE = InnoDB;"))
		{
			m_szLastError	= query.lastError().text();
			m_db.close();
			m_bConnected	= false;
			return(m_bConnected);
		}

		if(!query.exec("INSERT INTO `kooky`.settings values ('version', '1.0');"))
		{
			m_szLastError	= query.lastError().text();
			m_db.close();
			m_bConnected	= false;
			return(m_bConnected);
		}

		if(!query.exec("CREATE TABLE `kooky`.ingredient_group ( `id` INT NOT NULL AUTO_INCREMENT, `name` VARCHAR(100) NOT NULL, PRIMARY KEY (id) )"))
		{
			m_szLastError	= query.lastError().text();
			m_db.close();
			m_bConnected	= false;
			return(m_bConnected);
		}

		if(!query.exec("CREATE TABLE `kooky`.ingredient ( `id` INT NOT NULL AUTO_INCREMENT, `name` VARCHAR(100) NOT NULL, ingredient_group_id INT NOT NULL, PRIMARY KEY (id) )"))
		{
			m_szLastError	= query.lastError().text();
			m_db.close();
			m_bConnected	= false;
			return(m_bConnected);
		}

		if(!query.exec("CREATE TABLE `kooky`.ingredient_values ( `id` INT NOT NULL AUTO_INCREMENT, `ingredient_id` INT NOT NULL, `ingredient_number` INT NOT NULL, `value` DOUBLE, PRIMARY KEY (id), FOREIGN KEY(ingredient_id) REFERENCES ingredient(id) )"))
		{
			m_szLastError	= query.lastError().text();
			m_db.close();
			m_bConnected	= false;
			return(m_bConnected);
		}

		createGroup("default");
	}
	else
		m_szLastError	= m_db.lastError().text();

	return(m_bConnected);
}

bool cMySQLPlugin::open()
{
	m_bConnected	= m_db.open();

	if(m_bConnected)
	{
		QSqlQuery	query(m_db);
		if(!query.exec("SELECT * FROM settings"))
		{
			init();

			QSqlQuery	query1(m_db);
			if(!query1.exec("SELECT * FROM settings"))
			{
				m_szLastError	= query.lastError().text();
				m_db.close();
				m_bConnected	= false;
				return(m_bConnected);
			}
		}
	}
	else
		m_szLastError	= m_db.lastError().text();

	return(m_bConnected);
}

QString cMySQLPlugin::lastError()
{
	return(m_szLastError);
}

bool cMySQLPlugin::config()
{
	cConfigDialog	dlg;
	QSettings		settings;
	dlg.setHostname(settings.value(QString("plugins/%1/hostName").arg(pluginName())).toString());
	dlg.setDatabase(settings.value(QString("plugins/%1/databaseName").arg(pluginName())).toString());
	dlg.setUserName(settings.value(QString("plugins/%1/userName").arg(pluginName())).toString());
	dlg.setPassword(settings.value(QString("plugins/%1/password").arg(pluginName())).toString());

	if(dlg.exec() == QDialog::Accepted)
	{
		m_db.close();
		settings.setValue(QString("plugins/%1/hostName").arg(pluginName()), dlg.hostname());
		settings.setValue(QString("plugins/%1/databaseName").arg(pluginName()), dlg.database());
		settings.setValue(QString("plugins/%1/userName").arg(pluginName()), dlg.userName());
		settings.setValue(QString("plugins/%1/password").arg(pluginName()), dlg.password());

		m_db.setHostName(settings.value(QString("plugins/%1/hostName").arg(pluginName())).toString());
		m_db.setDatabaseName(settings.value(QString("plugins/%1/databaseName").arg(pluginName())).toString());
		m_db.setUserName(settings.value(QString("plugins/%1/userName").arg(pluginName())).toString());
		m_db.setPassword(settings.value(QString("plugins/%1/password").arg(pluginName())).toString());
		return(open());
	}

	return(m_bConnected);
}

qint32 cMySQLPlugin::exists(const QString& szName, const QString& szGroup)
{
	if(!m_bConnected)
		return(-1);

	qint32	iGroup	= existsGroup(szGroup);
	if(iGroup == -1)
		return(-1);

	QSqlQuery	query(m_db);

	if(!query.exec(QString("SELECT id FROM ingredient WHERE name = \"%1\" AND ingredient_group_id=%2").arg(szName).arg(iGroup)))
	{
		m_szLastError	= query.lastError().text();
		return(-1);
	}
	query.first();
	if(!query.isValid())
		return(-1);

	return(query.value(0).toInt());
}

qint32 cMySQLPlugin::existsGroup(const QString& szGroup)
{
	if(!m_bConnected)
		return(-1);

	QSqlQuery	query(m_db);

	if(!query.exec(QString("SELECT id FROM ingredient_group WHERE name = \"%1\"").arg(szGroup)))
	{
		m_szLastError	= query.lastError().text();
		return(-1);
	}
	query.first();
	if(!query.isValid())
		return(-1);

	return(query.value(0).toInt());
}

qint32 cMySQLPlugin::existsIngredient(qint32 id, qint16 ingredientNumber)
{
	if(!m_bConnected)
		return(-1);

	QSqlQuery	query(m_db);

	if(!query.exec(QString("SELECT id FROM ingredient_values WHERE ingredient_id = %1 AND ingredient_number = %2").arg(id).arg(ingredientNumber)))
	{
		m_szLastError	= query.lastError().text();
		return(-1);
	}
	query.first();
	if(!query.isValid())
		return(-1);

	return(query.value(0).toInt());
}

qint32 cMySQLPlugin::create(const QString& szName, const QString& szGroup)
{
	if(!m_bConnected)
		return(-1);

	qint32	iGroup	= existsGroup(szGroup);
	if(iGroup == -1)
		return(-1);

	if(exists(szName, szGroup) != -1)
		return(-1);

	QSqlQuery	query(m_db);
	if(!query.exec(QString("INSERT INTO ingredient (name, ingredient_group_id) VALUES (\"%1\", %2)").arg(szName).arg(iGroup)))
	{
		m_szLastError	= query.lastError().text();
		return(-1);
	}
	return(exists(szName, szGroup));
}

qint32 cMySQLPlugin::createGroup(const QString& szGroup)
{
	if(!m_bConnected)
		return(-1);

	if(existsGroup(szGroup) != -1)
		return(-1);

	QSqlQuery	query(m_db);
	if(!query.exec(QString("INSERT INTO ingredient_group (name) VALUES (\"%1\")").arg(szGroup)))
	{
		m_szLastError	= query.lastError().text();
		return(-1);
	}
	return(existsGroup(szGroup));
}

bool cMySQLPlugin::set(qint32 id, qint16 ingredientNumber, qreal value)
{
	if(!m_bConnected)
		return(-1);

	if(existsIngredient(id, ingredientNumber) != -1)
		return(-1);

	QSqlQuery	query(m_db);
	if(!query.exec(QString("INSERT INTO ingredient_values (ingredient_id, ingredient_number, value) VALUES (%1, %2, %3)").arg(id).arg(ingredientNumber).arg(value)))
	{
		m_szLastError	= query.lastError().text();
		return(-1);
	}
	return(existsIngredient(id, ingredientNumber));
}

INGREDIENT_LIST cMySQLPlugin::ingredients()
{
	if(!m_bConnected)
		return(INGREDIENT_LIST());

	QSqlQuery	query(m_db);
	if(!query.exec(QString("SELECT ingredient.name, ingredient_group.name FROM ingredient, ingredient_group WHERE ingredient.ingredient_group_id = ingredient_group.id ORDER BY ingredient_group.name, ingredient.name")))
	{
		m_szLastError	= query.lastError().text();
		return(INGREDIENT_LIST());
	}

	INGREDIENT_LIST	list;
	while(query.next())
	{
		INGREDIENT	i;
		i.szGroup		= query.value(1).toString();
		i.szIngredient	= query.value(0).toString();
		list.append(i);
	}

	return(list);
}

QStringList cMySQLPlugin::groups()
{
	if(!m_bConnected)
		return(QStringList());

	QSqlQuery	query(m_db);
	if(!query.exec(QString("SELECT name FROM ingredient_group ORDER BY name")))
	{
		m_szLastError	= query.lastError().text();
		return(QStringList());
	}

	QStringList	list;
	while(query.next())
		list.append(query.value(0).toString());

	return(list);
}

bool cMySQLPlugin::beginTransaction()
{
	if(!m_bConnected)
		return(false);
	QSqlQuery query(m_db);
	return(query.exec("SET autocommit = 0"));
}

bool cMySQLPlugin::endTransaction()
{
	if(!m_bConnected)
		return(false);
	QSqlQuery query(m_db);
	query.exec("SET autocommit = 1");
	return(m_db.commit());
}

QString cMySQLPlugin::author()
{
	return("Herwig Birke");
}

QString cMySQLPlugin::company()
{
	return("WIN-DESIGN");
}

QString cMySQLPlugin::link()
{
	return("http://www.windesign.at");
}

QString cMySQLPlugin::description()
{
	return("This is the description.");
}

QString cMySQLPlugin::copyright()
{
	return("This is the copyright.");
}
